### PR TITLE
Add unit test and feature

### PR DESCRIPTION
**Pull request by BanOrange**
- **test:新增了IO包和Demo包的测试用例**
- **test:Add some test cases for IO and Demo**
- **Feat:Configuration file injection**
- **fix:将JSON转化功能做了修改**
- **test:就JSON转化为配置的功能，撰写了测试类**
- **add SuitApplicationTest and SuitHostTests**

### CODE REVIEW

**Comment by BanOrange** [comment]
由于是多人协作，所以多个commit混在一起了，最后一个commit中两个都包含了

#增加了配置文件注入的功能

修改思路：由于我对C#确实不是很熟悉，同时没看懂C#版的代码中对于IConfiguration的示例。所以这里我是按照我的理解进行的配置文件注入。JsonConfiguration类可以使用parse()方法来解析一个用于配置SuitHost的Json文件，方法签名如下：

```
/**
 * return a SuitConfiguration from a Json File for the creation of SuitHost
 * @param JsonFilePath the path of the Json File,must be vaild
 * @return A SuitConfiguration Object,and if the file has wrong format,it will be a default CommonSuitConfiguration
 */
public SuitConfiguration parse(String JsonFilePath);
```

通过fastJSON库来解析JSON文件，如果发现有相关的字段而且字段值正确,就根据字段值来配置对应内容并最终返回一个Configuration以供用户生成SuitHost

示例文件

```
example.json

{
  "ColorSetting": {          //配置颜色的设置，其中颜色类型为ColorSetting的字段，颜色必须在ConsoleColor中出现
    "AllOkColor": "Blue",
    "ErrorColor": "Yellow"
  },
  "PromptServer": "Common", //配置PromptServer类型，仅支持Common或者PowerLineThemed
  "IOServer": {             //配置IOServer的相关流，不写或者System.xxx都表示使用标准流，此处input的value为“H”表示
    "input": "H",           //IOServer会使用一个新建的输入流，输入流中包括一个"H"字符
  	                        //并不建议改动IOServer的输入流
  },
  "Logger": {
    "AbsolutionPath": "D:/log4j.xml",      //配置Logger使用的配置文件绝对路径，优先级更高
    "RelativePath": "./log4j.xml"          //配置Logger使用的配置文件相对路径，优先级较低
  }
}
```

使用示例：

```
public static void main(String[] args) throws Exception {
    new SuitHost(Client.class,
            new JsonConfiguration().parse("src/main/resources/example.json")).Run();
}                                        
```

修改文件：

新增interface IConfiguration

新增JsonConfiguration.java

删除ColorSetting中各字段的final关键字





### **Add configuration file injection feature**

Modification idea: Since I am not very familiar with C# and did not understand the example of IConfiguration in the C# version. Therefore, I am making the configuration file injection according to my understanding. The JsonConfiguration class can use the parse() method to parse a JSON file used to configure SuitHost. The method signature is as follows:

```
/**
 * return a SuitConfiguration from a Json File for the creation of SuitHost
 * @param JsonFilePath the path of the Json File,must be vaild
 * @return A SuitConfiguration Object,and if the file has wrong format,it will be a default CommonSuitConfiguration
 */
public SuitConfiguration parse(String JsonFilePath);
```

Using the fastJSON library to parse a JSON file, if relevant fields are found and their values are correct, configure the corresponding content based on the field values and finally return a Configuration for users to generate SuitHost.



example JSON file

```
{
  "ColorSetting": {          // Configuration for color settings, where the color type is a field of                                      // ColorSetting, and the color must appear in ConsoleColor
    "AllOkColor": "Blue",
    "ErrorColor": "Yellow"
  },
  "PromptServer": "Common", // Configuration for PromptServer type, only supports Common or PowerLineThemed
  "IOServer": {             // Configuration for IOServer-related streams, empty or System.xxx means                               // using standard streams, here the value of input "H" indicates
    "input": "H",           // IOServer will use a newly created input stream, which includes a character                               // "H"
                            // It is not recommended to change the input and output streams of IOServer
  },
  "Logger": {
    "AbsolutionPath": "D:/log4j.xml",      // Configuration for the absolute path of the Logger                                                        // configuration file, higher priority
    "RelativePath": "./log4j.xml"          // Configuration for the relative path of the Logger                                                        // configuration file, lower priority
  }
}
```

A example to use your JSON file in Client.java

```
public static void main(String[] args) throws Exception {
    new SuitHost(Client.class,
            new JsonConfiguration().parse("src/main/resources/example.json")).Run();
}
```

Modifications of the files:

- Add  interface IConfiguration
- Add  JsonConfiguration.java
- Remove the final keyword from each field in ColorSetting



debug了部分Demo包中部分错误的测试用例，新增了更多的测试用例

Debug some error test cases in the demo package and added more test cases.

File: pom.xml [hunk 1]
```diff
@@ -107,6 +108,25 @@
             <scope>test</scope>
         </dependency>
 
+        <!--fafastjson-->
+        <dependency>
```
Comment (FerdinandSu) [inline pom.xml:112]: 能不能换成其他的json库，fastjson的名声不太好

File: src/main/java/ReFreSH/JMobileSuit/IO/ColorSetting.java [hunk 2]
```diff
@@ -10,31 +10,31 @@ public class ColorSetting {
     /**
      * Default color. For OutputType.Default
      */
-    public final ConsoleColor DefaultColor = ConsoleColor.White;
+    public ConsoleColor DefaultColor = ConsoleColor.White;
```
Comment (FerdinandSu) [inline src/main/java/ReFreSH/JMobileSuit/IO/ColorSetting.java:13]: 这个地方不需要使用配置，请回滚它

File: src/main/java/ReFreSH/JMobileSuit/IConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package ReFreSH.JMobileSuit;
+
+import ReFreSH.JMobileSuit.ObjectModel.SuitConfigurator;
+
+public interface IConfiguration {
+    /**
```
Comment (FerdinandSu) [inline src/main/java/ReFreSH/JMobileSuit/IConfiguration.java:6]: 这里应该是静态工厂方法public static IConfiguration parse(String JsonFilePath);
同时，IConfiguration应该是一个自递归的字典，See https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.configuration.iconfiguration?view=net-9.0-pp

File: src/main/java/ReFreSH/JMobileSuit/JsonConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+     * @param JsonFilePath the path of the Json File,must be valid
+     * @return The SuitConfiguration derived from the transformation.
+     */
+    @Override
+    public SuitConfiguration parse(String JsonFilePath) {
+        Logger logger = LogManager.getLogger(JsonConfiguration.class);
+        // Retrieve the JSON string from the file.
+        StringBuilder jsonContent = new StringBuilder();
+        try (BufferedReader reader = new BufferedReader(new FileReader(JsonFilePath))) {
+            String line;
```
Comment (FerdinandSu) [inline src/main/java/ReFreSH/JMobileSuit/JsonConfiguration.java:43]: IConfiguration并非是想硬编码成框架的配置，而是作为应用可以读取的配置

**Review by FerdinandSu — CHANGES_REQUESTED** [review]

**Pull request closed without merging**